### ISSUE TITLE
Effectful Event Handlers

### ISSUE THREAD

**Issue by mike-thompson-day8** [Day8/re-frame#174]
Side-effecting event handlers - how to do them better?

I'm seeking comment and feedback on this (incomplete) document:
https://github.com/Day8/re-frame/wiki/Effectful-Event-Handlers

It presents as a tutorial, but it is really a request for feedback and comments. 

#### Credit

We're discussing this because @metasoarous and @juhovh [recognised what was possible](https://github.com/Day8/re-frame/issues/86). I'm very glad they pushed the point. Ultimate credit should go to @qwtel for [his terrific blog post](http://takentech.com/side-effects-get-out-of-my-yard) earlier this year. He described the problem and gave the solution. Thankyou.

**Comment by ivanreese** [comment]
Like it a lot so far! Questions:

1) Under "Implementing `io`" we have side effect handlers, but pure isn't one of them. Any reason why pure wouldn't be turned into an effect handler?

2) How might one swap out side effect handlers at runtime? This feels a bit like kicking a beehive, but I'm surrounded by artists so I need to ask these sorts of questions because they'll try anything dammit.

3) Under "Swappable" you talk about switching instructions into noops. There are cases (eg: doing http while debugging) where you might just want to change the way the side effect is applied (eg: logging instead of triggering ajax). Currently, it looks like side effect handlers "claim" their key in the world map, so registering multiple handlers for the same key would be fraught, no?

**Comment by ds300** [comment]
Proposal: don't put the effects in middleware.

Rather, have side-effecting listeners hanging off your `world` which react to state changes to trigger effects. Let me illustrate how this would work in practice (n.b. I haven't run this code):

```clojure
;; so we augment the world with information about effects, like before
(def world (reagent/atom {:effects {:unhandled (sorted-map)
                                    :next-id 0}
                          :db {:whatevs true}}))

;; except rather than dealing with the effects in middleware, we just
;; throw them into a queue to be dealt with by someone else
(re-frame/register-handler
  :new-effect-created
  (fn [world [_ effect-descriptor]]
    (let [next-effect-id (get-in world [:effects :next-id])]
      (-> world
        (update-in [:effects :unhandled] assoc next-effect-id effect-descriptor)
        (update-in [:effects :next-id] inc)))))

;; and when that someone deals with them, the effect goes away unceremoniously
(re-frame/register-handler
  :effect-handled
  (fn [world [_ effect-id]]
    (update-in world [:effects :unhandled] dissoc effect-id)))

;; but where do the someones come from? Just like React, those someones
;; are frameworkey code which listens for state changes and does side
;; effects according to those state changes. Let's do a toy someone for
;; handling POST requests

;; but first let's "submit" a post request so we know what kind of
;; data we're dealing with here)
(dispatch [:new-effect-created
           {:type :post-request
            :data {:url "/users/new"
                   :on-success [:new-user-created]
                   :on-error [:new-user-request-failed]
                   :data {:name "The Pope" :age 139}}}])

;; ok so the frameworky someone listens for changes in the unhandled effects
(def unhandled-post-reqs (reagent/reaction
                           (->> @world
                            :effects
                            :unhandled
                            (filter #(= :post-request (:type (second %)))))))
                
;; and needs to maintain local state for keeping track of actual real-life requests
(def active-post-reqs (atom {}))


(reagent/track!
  (fn []
    ;; of course don't do actual post requests when replaying or debugging
    (when-not *replay-or-debug-mode-or-something-else-like-that*
      ;; find out which unhandled-post-reqs we're not already in the process
      ;; of handling, and start handling them
      (let [current-active-reqs @active-post-reqs
            new-active-reqs (reduce (fn [new-reqs [id effect]]
                                      (assoc new-reqs
                                             id
                                             (or (current-active-reqs id)
                                                 (make-new-xhr (:data effect) id))))
                                    {}
                                    @unhandled-post-reqs)]
        ;; cancel any reqs that were handled some other way or are no longer needed
        (doseq [[id xhr] current-active-reqs]
          (when-not (new-active-reqs id)
            (cancel xhr)))
        (reset! active-reqs new-active-reqs)))))

(defn make-new-xhr [{:keys [url, data, on-success, on-error]} id]
  (let [xhr (XMLHttpRequest.)
        success-handler (fn []
                          (dispatch [:effect-handled id])
                          (dispatch (conj on-success (.-response xhr))))
        error-handler (fn []
                        (dispatch [:effect-handled id])
                        (dispatch (conj on-error (.-response xhr))))]
    (.open xhr "POST" url)
    (.addEventListener "load" success-handler)
    (.addEventListener "error" error-handler)
    (.addEventListener "timeout" error-handler)
    (js/setTimeout #(.send xhr (json-serialize data)))
    xhr))

;; of course in reality you wouldn't be dispatching :new-effect-created from
;; your app code, but rather you'd do soemthing like this:

(re-frame/register-handler
  :new-user-form-submitted
  (fn [world [_ name age]]
    (add-new-effect world
                    (post-request {:url "/users/new"
                                   :on-success [:new-user-created]
                                   :on-error [:new-user-request-failed]
                                   :data {:name name :age age}}))))

;; where add-new-effect and post-request do the obvious things,

;; and then you could just be like
(dispatch [:new-user-form-submitted "The Pope" 139])
```

Makes sense? This way there is no synchronous chain between 'declaring' effects and having them executed, and therefore no need to treat effects as special-needs data. We can treats effects as normal application state to be reacted to, just like your garden variety virtual dom trees and whatnot.

From the event-handler-author's perspective it's the same idea as in your middleware proposal, but I think there's extra simplicity in actually keeping the effect descriptors in the application state and explicitly saying when they've been dealt with. It more closely models what's actually happening.

Another potential benefit is that this keeps all the data about effects within the domain of the app author, i.e. there's nothing stopping you from deriving some view based on the content of the unhandled effects queue. Want to know how many in-flight requests there are? Sure thing buddy!

Also, this makes it simple to add new effect types at runtime, which would presumably be somewhat awkward if re-frame had to know about all the effect types in order to dispatch effect descriptors to the appropriate handler.

**Comment by arichiardi** [comment]
The above is very very nice and I kind of agree that we'd a more general way to handle side effects. Again talking about Onyx, [this](https://youtu.be/kP8wImz-x4w) is a very nice video to watch and ponder about, especially starting from: https://youtu.be/kP8wImz-x4w?t=1948.

What is so special about is that it encapsulates side-effectful functions in data structures, as in Mike's approach, but it is not entirely tied with a handler execution point. Now, we are kind of tied anyways to it as we are running an event queue, but potentially we just need a "trigger" in order to execute the effect. Can this trigger happen only during the event dequeing or at any time? This needs some hammock time maybe.

Ok, I guess I am vouching the same proposal as @ds300 all in all. Mike, good job really, we are going in the right direction :+1:

**Comment by danielcompton** [comment]
From Kris Jenkin's excellent [What is Functional Programming?](http://blog.jenkster.com/2015/12/what-is-functional-programming.html)

> So the second piece of code has hidden inputs and outputs. It requires things, and causes things, but you could never guess what just by looking at the API.
> 
> These hidden inputs and outputs have an official name: "side-effects". There are many kinds of side-effects, but they all come together under the same concept: "when we call this function, what does it need that isn't in the argument list, and what does it do that isn't part of the return value?"
> 
> (Actually I think we need two terms: "side-effects" for the hidden outputs, and "side-causes" for the hidden inputs. For most of the rest of this post I'll use "side-effects" for brevity, but I'm definitely talking about side-causes too. I'm talking about all hidden inputs and outputs.)

This proposal is addressing the side effects side of handlers very well, but I think we should also consider side-causes i.e. inputs to world. Should handlers like local storage provide the contents of local storage to every function? Should it be lazily evaluated if so? What about side-causes like the current time?

**Comment by metasoarous** [comment]
Cheers! Great ideas here!

I very much agree with @arichiardi that the execution point should be as decoupled as possible. Different applications will need different transactional/coordination semantics for different situations, so this should be flexible. This separation could also simplify the problem of how to get noop, log, mock, swap out effect functions (for you crazy artists), etc, as well.

What if something like a transaction report was a more standard part of the re-frame ethos? Perhaps this could lead to a more general approach for responding to information associated with a particular transaction (or transactions), leaving the trigger semantics as a separate concern. I don't know what this might look like though, or how you'd pass that data along.

Of course, DataScript and Datomic _have_ transaction reports, and _transaction metadata_ to boot. And of course, in Datomic at least, this data is _persisted_ (it can be in DataScript as well). So this could be an attractive target in our case. But what makes this rather easy is that we make strong assumptions about the shape of your db: It's DataScript. But for Re-frame, it seems you'd have to make at least _some_ assumptions about the shape of the db in order to have a general purpose solution, and I don't know how varied folks get in how they structure their dbs. Unless Re-frame did something that really was agnostic like 

Final thing to think about with all this: Imagine you have a rather complicated event handler which calls out to several other event handlers in turn. If several of those sub-handlers have side effects, what should happen if one of them fails? How much control should users have over this? What behavior is least surprise?

@danielcompton Great point! I think the one thing we'd _always_ want to be sure of is that throughout the course of a transaction, any value from a side-cause be consistent throughout the course of that transaction. But this leaves the semantics of _how_ those snapshots are populated (lazily, greedily, mixed for different inputs...) as a separate concern. One way to do this is to force all side-causing data sources to push their data through the dispatch function. That way the value used for something is just the last value reflected in the data base. There might be other ways to do this that could have advantages but I can't really think of any. The one more challenging scenario I can think of is where you would have some set of input signals that you wanted to sample at "approximately the same time" and submit as a single transaction. If you just arbitrarily feed in separate event streams you could get updates that are based on data that is mismatched by some interval of time on the order of the sampling rate. So I think all we could really use is some tooling around how we sample things in scenarios such as this.

**Comment by ds300** [comment]
@danielcompton 
The term 'side-cause' seems whack to me. If I understand Kris Jenkins correctly, he is using the term to simply mean *reading some global mutable state*. There is no element of causality involved.

Anyway I'm not sure that the act of reading global mutable state needs to be formalised or facilitated in any way. Especially not in a single-threaded environment. We can just dereference it in our 'external event' handlers and `dispatch` it into re-frame as immutable domain event data. We might even modify it before reading it, as in random number generation. If some data signal, like the current time, is not important to your domain model but important for triggering side effects like view rendering, then you can just shove it into a separate ratom rather than dispatching it into your db.

@metasoarous 
Maybe a transaction report in re-frame would just be a diff?

**Comment by slorber** [comment]
@mike-thompson-day8 I don't speak Clojure so can't easily read that much your code, but if you understand Elm or JS please take a look at these links that might interest you.

- https://github.com/slorber/scalable-frontend-with-elm-or-redux
- http://stackoverflow.com/questions/34570758/why-do-we-need-middleware-for-async-flow-in-redux/34623840#34623840

Redux-saga or Elm permit to yield an "effect description" (easy to test), that will be run against an effect interpreter. Note that all side-effects are not always technical (like async requests) but sometimes it's more business related (on event1, fire event2)

**Comment by samroberton** [comment]
@mike-thompson-day8 :  In case it's of interest, here's what I do.  (Bearing in mind of course that this is within my own state machines / statecharts setup, not using re-frame.)

I require all transition functions (roughly equivalent to re-frame handlers) to be pure.  The `db` map which they get passed contains a queue under `:outputs`.  Any transition function which wants to cause a side effect `conj`s something (can be whatever) onto `:outputs`.  I have a watch on the DB atom which just looks for non-empty `:outputs`, and invokes an application-supplied `output-handler`.  This `output-handler` gets invoked with each item on the `:outputs` queue.  In practice, the "something" you put into `:outputs` is a map with an `:id` and arbitrary other data, and the `output-handler` is a `defmulti` which dispatches on `:id`.  This means all my side-effects are contained within the `defmethod`s of that multimethod.  I was surprised by how few there actually are.

One advantage of this is that for testing purposes, you can just supply a different `output-handler` method.  Or alternatively, you can remove the watch, so that your `:outputs` remain on the queue, and in your tests, you manually pop them off after asserting that they contain the right thing, then you manually dispatch eg the AJAX response that you want your test to handle.  (I actually find in my tests I tend to run things until the app is in the state I want, then remove the watch so the test can manually handle the output and any subsequent resulting input.)

As it happens, I've been toying with renaming my `db` to `world`.  Wondering about renaming `outputs` to `actions`, too, but `io` is an interesting alternative.

**Comment by keeth** [comment]
Not sure if this is the right place for this comment, but I've just started my first re-frame app after working extensively with redux and redux-saga, and I find myself wanting a saga-like pattern in re-frame.

When adding effects to individual handlers in re-frame I immediately run into a kind of action/handler spaghetti where it is hard to see the overall flow because it is spread out across multiple handlers.. the main place where I use this is in app boot/login, where I 

* :load-config - loads a config json, on success dispatch :attempt-token-auth
* :attempt-token-auth - attempts to login to firebase with token from local storage, on success dispatch :get-data
* :get-data - load some initial data from the database
etc.
etc.

There are also dispatching/handling of various error conditions at each stage.

This flow cuts across different areas of my app, and so to follow the flow you need to jump around in the code to different handlers, which are in different files.

The big win from redux-saga is that you can centralize complicated interactions and biz logic into unified generators that yield effects to the saga middleware.  The key similarity to what's being discussed here is yielding effects as data to be executed by middleware. The key difference is that rather than reacting to a single action type, a saga can wait for any action (or wait for a race between multiple actions), can dispatch actions, wait for the result of a promise, etc, all using a synchronous style which avoids callback hell.

I could see this being done in re-frame with the help of core.async, where the a go routine yields the effects to a channel to be executed, and can pause waiting for certain things to happen.

```
(defn main-flow [ch]
  ; wait for app to start (:boot-app action)
  (<! (>! ch [:await :boot-app]))
  (go-loop []
           ; yield an http effect
           (let [[err config] (<! (>! ch [:http :get "./config.json"]))]
             (if err 
               (>! ch [:dispatch :config-error err])
               (do 
                 ; yield an action dispatch effect
                 (>! ch [:dispatch :config-success config])
                 ; get a value from local storage
                 (let [token (<! (>! ch [:local-storage :get "token"]))]
                   (if token
                     ; dispatch :init-db action
                     (>! ch [:dispatch :init-db config token])
                     (do 
                       ; prompt the user for login, and wait for them to submit the login form
                       (>! ch [:dispatch :prompt-for-login])
                       (<! (>! ch [:await :login])))
                     ; handle login, etc, etc
                     )))))
    recur)
  )
```

Similar to redux-saga i could also see a race-like construct, to wait for the first of several events to happen, e.g. 

```
(let [[winner payload] (<! (>! ch [:race {:some-action-happened [:await :some-action] 
                                          :timed-out [:timeout 3000]}]))]
  (cond (= :some-action-happened winner) (do-something payload)
        (= :timed-out winner) (>! ch [:dispatch :timeout-error])))
```

This is what's been rattling around in my head the last couple days.. curious to hear what others think.

**Comment by aft-luke** [comment]
@keeth I agree wholeheartedly about the power and convenience of a saga-like centralized "business behavior processor". I would be interested to hear from others more experienced in re-frame if this is truly a concern, or if the problem doesn't arise often enough to warrant a special construct?

I also came over from the redux/saga world. I believe redux *without* sagas is a very, shall we say, misleading pattern. It's all roses and butterflies until your project is significantly large and you need to make a non-trivial extension / change. It does not prove adaptable enough for my tastes, and while looking for a solution I stumbled on redux-saga. In the end I wondered if redux-saga is perhaps a bandaid (albeit a great bandaid, but a bandaid nonetheless) over a larger architectural flaw.

Reframe does not suffer from the same underlying issues as redux. That being said, the centralized nature of a single unit that processes complex business logic in a psuedo-sequential order, all while yielding effects as data, is a great idea. It really eases the flow of operations such as (as you mentioned) bootstrapping with auth/auth, large application transitions into deeply nested states, and if-error-otherwise use cases. 

I can see a saga-like pattern being most useful in reframe in one particular case-- some feature's handler affects change onto the db, and another completely independent feature elsewhere needs to respond not just from reading the change from the db, but by going and performing some async behavior. A "solution" here is to have the original handler take care of both cases, but it just feels icky to have feature foo suddenly understand feature bar's needs. To avoid this, one can create a new higher-level feature that sits on top of foo and bar, understanding both their needs, but such a change requires shifting around handlers, perhaps rearranging the shape the db, and increasing the number of props/dispatch callbacks passed down through the view tree.

**Comment by keeth** [comment]
@aft-luke yep I consider redux-saga an essential pillar of a redux app, there's just no other good story for higher level business logic - I see people tend to either put it all into the view layer, which sucks, or use something like thunk middleware, which is a partial solution but loses the expressiveness of effects-as-data.  Agree re-frame is not as limited as redux in terms of what action handlers do, but where side-effects are concerned I'm finding the same problem in re-frame of basically a scattered pub-sub model where complex flows are difficult to follow.

**Comment by mike-thompson-day8** [comment]
My incomplete WIP thoughts [are captured here](https://gist.github.com/mike-thompson-day8/592ba996321f5887a7ff819bcce57a1a)

**Comment by ds300** [comment]
Oh man, I like this idea a lot 👌 . Isn't it more generally applicable? Seems like it can describe any old async flow, not just app boot sequences.

**Comment by mike-thompson-day8** [comment]
@ds300  Yes, the approach should be general. The interesting question is how to "package it up" so that it is easy and intuitive to use.

The  core.async (or promises) approach is to create "Task Flow" in code.  A state machine is being created in code via the use of go loops and puts and gets (or `forks` and `

I would prefer to come up with a solution in which the required "Task Flow"  is described in data.  That's where I'm angling.  I'm sure there's a neat solution lurking nearby.

**Comment by keeth** [comment]
Some thoughts:

* Agree that a general approach makes sense.. there are many applications for this kind of system beyond boot.  
* Agree that both of these approaches (core.async vs data) are describing task flows or state machines by different means.  I understand not wanting to force core.async on users of re-frame.
* I like that in your proposal the actual details of the work does not leak into the coordinator - it just cares about tasks and events.  That separation is nice in that the coordinator is wiring together these various handlers without knowing too much about them.
* I'm glad I now know about the *add-post-event-callback* feature.  Seems there's no reason why someone could not implement a saga-like thing as a separate module.

**Comment by danielcompton** [comment]
> @ds300 "Isn't it more generally applicable?"

I had the exact same question when I first heard this approach described, and was happy to see that it was usable anywhere.

**Comment by mike-thompson-day8** [comment]
For anyone following along, I've now implemented the idea sketched out above:  
https://github.com/Day8/re-frame-async-flow-fx

**Comment by ds300** [comment]
🎉 

A couple of questions:

1. Have you thought any more about how to explicitly thread data through the flow? I have a little, but don't have anything to show for it yet. Feels like there's a really neat solution in there somewhere...

2. Regarding the implementation of the flow event handler: I'd have taken a slightly different approach, one which perhaps you've already dismissed. So i'd be interested in why if that's the case. 

  > the event handler keeps track of what events have occurred, and what tasks have already been started. It keeps this state at the path nominated in :db-path.

  > the event handler uses your flow specification and the state it internally maintains to work out how it should respond to each newly forwarded event.

  Instead of separately keeping track of which events have been seen and which tasks have been started, I'd just update the user's flow spec. e.g.

  Maybe the world state looks like this

  ```clojure
  {:db {...app state}
    :flow-rules
     [{:when :seen? :events #{:success-X :success-Y}  :dispatch [:do-Z]}
      ... more rules]}
  ```

  Then when `:success-X` happens the world state looks like this:

  ```clojure
  {:db {...app state}
    :flow-rules
     [{:when :seen? :events #{:success-Y}  :dispatch [:do-Z]}
      ... more rules]}
  ```

  Then when `:success-Y` happens `:do-Z` is dispatched and the world state looks like this:

  ```clojure
  {:db {...app state}
    :flow-rules
     [... more rules]}
  ```

  I can imagine wanting to keep the user's full spec around to support things like automatic retries, but not sure whether you're thinking along those lines.

**Comment by mike-thompson-day8** [comment]
> Have you thought any more about how to explicitly thread data through the flow? I have a little, but don't have anything to show for it yet. Feels like there's a really neat solution in there somewhere...

One of the central principals in re-frame is that all the data (as much as possible) is in app-db.  So I haven't thought much about state being keep outside of it.

> Regarding the implementation of the flow event handler: I'd have taken a slightly different approach ...

Yes, I can see where you are going with that, but it is probably 6 of one and half a dozen of another.  Nothing compelling either way, so I'll stick with what is now written and tested.

**Comment by ds300** [comment]
> One of the central principles in re-frame is that all the data (as much as possible) is in app-db.

I see no reason to violate that. The scheme I'm thinking of would just allow one to store data in the flow spec (or flow handler state) to pass along when dispatching dependent events.

```clojure
{:db {...app state}
  :flow-rules
   [{:when :seen?
     :events #{:success-X :success-Y}
     :dispatch [[:do-Z 'data-X 'data-A] [:do-Q 'data-Y 'data-X]]}
    ... more rules]}
  :flow-data {'data-A {... some data}}
```

But this is an extremely rough sketch of an idea. No idea if it would actually be practical.

**Comment by pbkhrv** [comment]
I like the idea of effectful handlers. Been reading this thread and some wiki pages. I'd like to contribute some real-world feedback from a customer-facing re-frame app, that's been running in production for > 6 months, using re-frame v0.6:
- I have 150+ handlers
- 91% of handlers change the app db (46% *ONLY* change the db, the rest combine several side effects)
- 23% of handlers call a RESTful API on the server side (using core.async+http) and then dispatch based on success or failure (inside of async's `go` block)
- 15% of handlers call native JavaScript functions (because I, unfortunately, have 2 components implemented in JS that haven't been reimplemented in CLJS yet)
- 20% of handlers `dispatch` something
- As for "exotic" side effects, I have handlers that
  - save into browser's localstate
  - use `secretary` to navigate to different subpages of the SPA (4%)

Overall, about a third of the re-frame handlers that I have are part of a "flow" that could be described using something like `re-frame-async-flow-fx` mentioned above

## The thing about API on the server

Our app is very heavy on calling server-side API (as I suspect is the case for majority of re-frame apps out there). If I were to start using flow-based patterns, I'd create a separate :my-api fx and use that to hide all of the API-related mess inside, instead of using something like `re-frame-http-fx`. It would be a way to move one step up the abstraction ladder - can mock the whole API out etc etc.

Just my 2 cents.

**Comment by mike-thompson-day8** [comment]
Status:  Getting close to a release.  

Docs are developing here:  https://github.com/Day8/re-frame/tree/develop/docs

**Comment by mike-thompson-day8** [comment]
Released as part of v0.8.0
Docs available:  https://github.com/Day8/re-frame/tree/develop/docs
